src/IK/src/QPInverseKinematics.cpp

```diff
@@ -0,0 +1,405 @@
+        }
+        m_pimpl->tasks[taskName].weight = weight.value();
+    }
+
+    // take the right task
+    m_pimpl->tasks[taskName].task = task;
+    m_pimpl->tasks[taskName].priority = priority;
+
+    if (priority == 0)
+    {
+        if (isInequality)
+            m_pimpl->inequalityConstraints.push_back(m_pimpl->tasks[taskName]);
+        else
+            m_pimpl->equalityConstraints.push_back(m_pimpl->tasks[taskName]);
+    } else
+    {
+        m_pimpl->costs.push_back(m_pimpl->tasks[taskName]);
+    }
+
+    return true;
+}
+
+bool QPInverseKinematics::finalize(const System::VariablesHandler& handler)
+{
+    constexpr auto logPrefix = "[QPInverseKinematics::finalize]";
+
+    //
+    std::size_t numberOfConstraints = 0;
+    for (auto& [name, task] : m_pimpl->tasks)
+    {
+        if (!task.task->setVariableHandler(handler))
+        {
+            log()->error("{} Error while setting the variable handler in the task named {}, having "
+                         "the following description {}.",
+                         logPrefix,
+                         name,
+                         task.task->getDescription());
+            return false;
+        }
+
+        if (task.priority == 0)
+        {
+            numberOfConstraints += task.task->size();
+        }
+    }
+
+    m_pimpl->solver.data()->setNumberOfVariables(handler.getNumberOfVariables());
+    m_pimpl->solver.data()->setNumberOfConstraints(numberOfConstraints);
+
+    // resize matrices
+    m_pimpl->hessian.resize(handler.getNumberOfVariables(), handler.getNumberOfVariables());
+    m_pimpl->gradient.resize(handler.getNumberOfVariables());
+
+    m_pimpl->constraintMatrix.resize(numberOfConstraints, handler.getNumberOfVariables());
+    m_pimpl->upperBound.resize(numberOfConstraints);
+    m_pimpl->lowerBound.resize(numberOfConstraints);
+
+    if (!handler.getVariable(m_pimpl->robotVelocityVariable.name,
+                             m_pimpl->robotVelocityVariable))
+    {
+        log()->error("{} Error while retrieving the robot velocity variable.", logPrefix);
+        return false;
+    }
+
+    return true;
+}
+
+std::vector<std::string> QPInverseKinematics::getTasksName() const
+{
+    std::vector<std::string> tasksName;
+
+    for (const auto& [name, task] : m_pimpl->tasks)
+    {
+        tasksName.push_back(name);
+    }
+
+    return tasksName;
+}
+
+bool QPInverseKinematics::advance()
+{
+    constexpr auto logPrefix = "[QPInverseKinematics::advance]";
+
+    // when advance is called the previous solution is no more valid
+    m_pimpl->isValid = false;
+
+    if (!m_pimpl->isInitialized)
+    {
+        log()->error("{} Please call initialize() before advance().", logPrefix);
+        return false;
+    }
+
+    // update of all the tasks
+    for (auto& [name, task] : m_pimpl->tasks)
+    {
+        task.task->update();
+    }
+
+    // Compute the gradient and the hessian
+    m_pimpl->hessian.setZero();
+    m_pimpl->gradient.setZero();
```

Comment:
The hessian is given by the sum of the hessian (same for the gradient).

I think this is related to https://github.com/dic-iit/bipedal-locomotion-framework/pull/229#discussion_r590588096

Reply:
So if the entirety of the hessian and the gradient are recomputed every time, there is no need to initialize them to zero in the update method. I agree on initializing them to zero once, but then doing that at every control loop seems not necessary.

Reply:
Ah no, I see it now, sorry for the delay :grin: Since you are adding things, you need to initialize to something. You could have initialized using the first hessian, but in this way, it is cleaner. 

Ignore this comment